src/java.base/share/classes/com/sun/crypto/provider/ChaCha20Cipher.java

```diff
@@ -831,31 +840,34 @@ private void setInitialState() throws InvalidKeyException {
         }
     }
 
-    /**
-     * Using the current state and counter create the next set of keystream
-     * bytes.  This method will generate the next 512 bits of keystream and
-     * return it in the {@code keyStream} parameter.  Following the
-     * block function the counter will be incremented.
-     */
-    private void generateKeystream() {
-        chaCha20Block(startState, counter, keyStream);
-        counter++;
+    @ForceInline
+    private static int chaCha20Block(int[] initState, long counter,
+            byte[] result) {
+        if (initState.length != KS_SIZE_INTS || result.length != KS_MAX_LEN) {
+            throw new IllegalArgumentException(
+                    "Illegal state or keystream buffer length");
+        }
+
+        // Set the counter value before sending into the underlying
+        // private block method
+        initState[12] = (int)counter;
+        return _chaCha20Block(initState, result);
     }
 
     /**
      * Perform a full 20-round ChaCha20 transform on the initial state.
      *
-     * @param initState the starting state, not including the counter
-     *      value.
-     * @param counter the counter value to apply
+     * @param initState the starting state using the current counter value.
      * @param result  the array that will hold the result of the ChaCha20
      *      block function.
      *
-     * @note it is the caller's responsibility to ensure that the workState
-     * is sized the same as the initState, no checking is performed internally.
+     * @return the number of keystream bytes generated.  In a pure Java method
+     *      this will always be 64 bytes, but intrinsics that make use of
+     *      AVX2 or AVX512 registers may generate multiple blocks of keystream
+     *      in a single call and therefore may be a larger multiple of 64.
      */
-    private static void chaCha20Block(int[] initState, long counter,
-                                      byte[] result) {
+    @IntrinsicCandidate
+    private static int _chaCha20Block(int[] initState, byte[] result) {
```

Comment:
Seems like there are 2 major naming conventions for intrinsic helper methods: prepend "impl" (e.g, `CounterMode.implCrypt`) or append "0" (`GaloisCounterMode.implGCMCrypt0`). I'd prefer to see either one used here.

Reply:
Done.